blog/_posts/2021-11-04-docker-build-alternatives.md

```diff
@@ -0,0 +1,234 @@
+is for you.
+
+It's a drop-in replacement for `docker build`.
+
+When you run `kubectl build`, the CLI:
+
+- Deploys a Docker build server to your Kubernetes cluster.
+
+- Sends your Dockerfile and source code to the build server.
+
+- Runs the build and pushes to a registry.
+
+A few months ago, [Gaëtan Lehmann](https://github.com/glehmann) added an
+extension for Tilt that lets you use `kubectl build` for image-building.
+
+If you have this Tiltfile:
+
+```python
+docker_build(
+  'example-python-image', 
+  '.', 
+  build_args={'flask_env': 'development'},
+  live_update=[
+    sync('.', '/app'),
+  ])
+```
+
+You can replace it with:
+
+```python
+load('ext://kubectl_build', 'kubectl_build')
+kubectl_build(
+  'example-python-image', 
+  '.', 
+  build_args={'flask_env': 'development'},
+  live_update=[
+    sync('.', '/app'),
+  ])
+```
+
+The main difference is that you will now see a buildkit server in your cluster:
+
+```shell
+$ kubectl get deployment
+NAME             READY   UP-TO-DATE   AVAILABLE   AGE
+buildkit         1/1     1            1           70m
+example-python   1/1     1            1           7s
+```
+
+This can be nice if you have a powerful cluster and a 13" laptop!
+
+Read the full docs on the `kubectl_build` extension [here](https://github.com/tilt-dev/tilt-extensions/tree/master/kubectl_build).
+
+## Ko
+
+Maybe you're not happy with your Dockerfile.
+
+The Dockerfile view of the world is: to build an image, run some commands in a container,
+then save the results of those commands.
+
+The Ko view is: an image is a tarball. If you have a Mac, why do you need a
+Linux VM to run a container if you want a tarball? You already have good local tools
+for making tarballs.
+
+[Ko](https://github.com/google/ko#install) is an image builder aimed at Go
+apps. Go apps are particularly ill-suited for `docker build`. The Go toolchain is
+well-optimized to run locally and builds a single binary. Ko builds your Go app
+locally, then packs it into an image. No build containers or VMs needed!
+
+A while back, I wrote a small extension that lets you use `ko` in Tilt. Here's what it looks like:
+
+```
+load('ext://ko', 'ko_build')
+
+ko_build(
+  'example-go-image',
+  './',
+  deps=['.']
+)
+```
+
+The upside of `ko` is that it's super simple to get started.
+
+The downside is that if you want any features that can't be expressed in a
+static Go binary, you're going to have a difficult time. Not impossible, but
+difficult. This includes: static files on-disk, dynamic C libraries, or
+hot-reload.
+
+Read the full docs on the `ko` extension
+[here](https://github.com/tilt-dev/tilt-extensions/tree/master/ko).
+
+## Pack
+
+[`pack`](https://buildpacks.io/docs/tools/pack/) is also addressing people who
+are not happy maintaining Dockerfiles.
+
+But where Ko's philosophy states that you don't need a container to build an
+image, Buildpack takes the viewpoint that you need MORE containers.
```

Comment:
I would say the philosophy is that you shouldn't have to care how an image is built, like if a tool is using more or fewer containers to make that happen. Give `pack` your source code, it'll give you an image. What happens in between are details that you as an app dev don't need to manage or care about.

Reply:
haha, yes, this was a bit tongue-and-cheek, but you're right that i should fold in the actual philosophy and not the joke one